### PR TITLE
[RFC] twister: Extend with debug capabilities

### PR DESCRIPTION
**Pull request by golowanow**
updated 10/10/2023:
An extension to Twister Harness and Test Instance to facilitate triage/investigations/mocking of test cases running in batch mode (on CI) with attached GDB scripted process running in parallel with the test: 

* Introduce Agent class to serve to Harness class as a delegate for control functions separated from log output interpretation functions.
An optional Agent object facilitates general process controls when Harness needs to run an external programm like test data interpreter, filter, or test flow controller.
Agent can also serve as additional lightweight data provider either from the same DUT managed by Handler or from another physical device.
Currently Twister has two kinds of Harnesses: 'simple' harnesses (Test, Ztest, Gtest, Console) and 'heavy' harnesses (Robot and Pytest).  The Agent class should be helpful for refactoring 'heavy' harnesses for code unification and reuse.

* Introduce Gdbagent as a Harness Agent to control a GDB process which connects to the test application or the device under test (DUT) and participates in the Test Instance execution.
The Zephyr gdbstub debug feature, QEMU gdbstub, or any other GDB compatible backend might be used at the DUT.
Some possible use cases:
      * to collect more data from the DUT: registers, memory, stack trace, etc. while executing the test and without changing its binary image.
      * to introduce conditional breakpoints and changes in DUT behavior which are not a part of the test image, do fuzzing, chaos testing, etc.
      * quickly implement and deploy mock functions, e.g. for sensors.
      *  use GDB scripts from developer's investigations to reproduce rare issues during batch test execution on CI.
      * fault injection (#3559)

* Introduce optional debug harness at Twister TestSuite to allow configurations with two simultaneous control/data test flows.
The debug harness can be added to an existent test definition to set up a debug script execution (e.g. with GDB and GDB backend on the DUT) while the test application itself is running unchanged, so rare issues can be investigated in batch mode.
The second debug Harness allows to have combinations of Harnesses and Agents of different type, for example:
a 'heavy' Harness (like Pytest) might use its own Agent to control the test execution while the Debug Harness runs GDB Agent using Console harness or some other Pytest harness.

* Add GDBstub over serial test suite also as a test example for the proposed feature.

Supported fixes/changes:

* Add GDBSTUB_TRACE config option to extend GDB backend logging for remote commands received and to debug the GDB stub itself  when needed.

* (also in separate #63621) Fix the Twister Console harness ordered 'multi_line' and 'one_line' pattern matching to treat the ztest as failed if not all the expected patterns were found in the console output, but the ztest application itself reports 'PROJECT EXECUTION SUCCESSFUL'.

### CODE REVIEW

File: subsys/debug/Kconfig [hunk 1]
```diff
@@ -423,6 +423,13 @@ config GDBSTUB_BUF_SZ
 	  for GDB backend. This needs to be big enough to hold one
 	  full GDB packet at a time.
 
+config GDBSTUB_TRACE
+	bool "GDB backend extra logging"
+	default n
```
Comment (dcpleung) [inline subsys/debug/Kconfig:428]: Default is `n` so there is no need for this line.

**Comment by gchwier** [comment]
Can you check PR #63061 ?
Maybe could be easier to use pytest instead of adding new harness to twister?

**Comment by golowanow** [comment]
> Can you check PR #63061 ? Maybe could be easier to use pytest instead of adding new harness to twister?

@gchwier 
with the proposed Twister  debug harness is it possible to extend any twister test with additional GDB control flow using the same yaml syntax for test cases. This feature can be used to automate investigations e.g. of test cases with 'random' errors.

The pytest plugin with gdbstub  is useful for sure with pytest-oriented test collections or to automate some complex non-linear debugging scenario. I'm going to extend #63061 sample as well and compare its performance on bulk CI executions.

**Review by nashif — CHANGES_REQUESTED** [review]
This is duplicating functionality already proivided by pytest. Using console evaluation based on regex is something we want to avoid in tests, this primarily reserved for samples. It will be very difficult to maintain something like this cross architecture and extend tests.

| This feature can be used to automate investigations e.g. of test cases with 'random' errors.

such investigation need to be done outside of twister. We should not look at twister as a debugging tool.

Adding such usecase/feature specific harnesses, i.e. gdb_stub is not scalable and will introduce more complexity into twister, something we can avoid by using pytest. 

I suggest to move the gdb stub to become a test and enhance it instead of introducing a dedicated gdb stub harness that will be limited, i.e. you will not be able to simulate interactive  gdb sessions with that...

**Comment by golowanow** [comment]
@nashif, this proposal is to to extend debug capabilities while testing to investigate difficult issues while running tests on CI ('at scale').

It is currently related to the GDBstub Zephyr software backend, but the same 'debug harness' can be used with other debug agents.

It doesn't depend on pytest plugin for simplicity and to allow any existent test to be debugged if needed.

For example, if we have some test for HW sensors which sometimes/rarely fails and it is difficult to catch manually why.
We can quite simply add debug harness definition to that test and to collect more low level data while the test application runs almost as usual, at certain GDB watchpoints, without changing the application itself, just with GDB script and few optional regexp patterns to check the gdb script itself works properly.
Another related use case is to quickly implement a mock, e.g. for a sensor.
These use cases will reuse what developer has done to investigate a problem manually with gdb. It does not require someone to develop a new python test for that.

An interactive test is also possible to implement this way, as a GDB script with some if/else logic, but it is not a main purpose for this feature extension.

**Pull request closed without merging**